File: clonebot/progress.py
from __future__ import annotations

import html
import math
import time
from dataclasses import dataclass, field
from typing import Optional


def _fmt_bytes(num_bytes: float) -> str:
    units = ["B", "KB", "MB", "GB", "TB"]
    value = float(max(num_bytes, 0.0))
    for unit in units:
        if value < 1024.0 or unit == units[-1]:
            if unit == "B":
                return f"{int(value)} {unit}"
            return f"{value:.2f} {unit}"
        value /= 1024.0
    return f"{value:.2f} TB"


def _fmt_duration(seconds: float) -> str:
    if not math.isfinite(seconds) or seconds < 0:
        return "--"
    sec = int(seconds)
    h, rem = divmod(sec, 3600)
    m, s = divmod(rem, 60)
    if h > 0:
        return f"{h:02d}:{m:02d}:{s:02d}"
    return f"{m:02d}:{s:02d}"


def _clip(text: str, max_len: int = 80) -> str:
    text = (text or "").strip().replace("\n", " ")
    if len(text) <= max_len:
        return text
    return text[: max_len - 3] + "..."


@dataclass
class CloneProgress:
    task_name: str
    total_bytes: int = 0
    total_files: int = 0
    copied_bytes: int = 0
    copied_files: int = 0
    started_at: float = field(default_factory=time.time)
    status: str = "Initializing"

    @property
    def percentage(self) -> float:
        if self.total_bytes > 0:
            return min(100.0, (self.copied_bytes / self.total_bytes) * 100.0)
        if self.total_files > 0:
            return min(100.0, (self.copied_files / self.total_files) * 100.0)
        return 0.0

    @property
    def elapsed(self) -> float:
        return max(0.001, time.time() - self.started_at)

    @property
    def speed_bps(self) -> float:
        return self.copied_bytes / self.elapsed if self.copied_bytes > 0 else 0.0

    @property
    def eta_seconds(self) -> float:
        if self.total_bytes <= 0:
            return float("nan")
        remaining = max(0, self.total_bytes - self.copied_bytes)
        speed = self.speed_bps
        if speed <= 0:
            return float("nan")
        return remaining / speed

    def progress_bar(self, width: int = 12) -> str:
        pct = max(0.0, min(100.0, self.percentage))
        filled = int((pct / 100.0) * width)
        return "[" + ("▓" * filled) + ("░" * (width - filled)) + "]"

    def as_message(self) -> str:
        safe_name = html.escape(_clip(self.task_name, 90))
        total_size = _fmt_bytes(self.total_bytes) if self.total_bytes else "Unknown"
        safe_bar = html.escape(self.progress_bar())
        safe_done = html.escape(_fmt_bytes(self.copied_bytes))
        safe_total = html.escape(total_size)
        safe_speed = html.escape(f"{_fmt_bytes(self.speed_bps)}/s")
        safe_eta = html.escape(_fmt_duration(self.eta_seconds))

        return (
            "<u><b>CLONING</b></u>\n"
            f"<code>{safe_name}</code>\n"
            f"<code>{safe_bar} {self.percentage:.2f}%</code>\n"
            f"<code>{safe_done}</code> <code>/</code> <code>{safe_total}</code>\n"
            f"<code>↓{safe_speed} | ETA : {safe_eta}</code>"
        )

    def completion_message(self, output_name: str, output_url: str, mime_type: str) -> str:
        safe_name = html.escape(output_name)
        safe_size = html.escape(_fmt_bytes(self.copied_bytes))
        safe_url = html.escape(output_url, quote=True)
        return (
            "<code>CLONED:</code>\n"
            f"<code>{safe_name}</code>\n"
            f"<code>{safe_size} • </code>"
            f"<a href=\"{safe_url}\"><b>DL</b></a>"
        )


class ThrottledReporter:
    def __init__(self, min_interval_seconds: float = 2.5) -> None:
        self.min_interval_seconds = min_interval_seconds
        self._last_update_at = 0.0

    def should_update(self, force: bool = False) -> bool:
        now = time.time()
        if force or now - self._last_update_at >= self.min_interval_seconds:
            self._last_update_at = now
            return True
        return False

File: internal/progress/progress.go
// Package progress renders live clone progress as Telegram HTML messages.
package progress

import (
	"fmt"
	"html"
	"math"
	"strconv"
	"strings"
	"time"
)

// FormatBytes renders a byte count using binary units, rounded to two decimals
// with trailing zeros trimmed: "1.5 KB", "1.0 MB", "1.25 GB".
func FormatBytes(numBytes float64) string {
	units := []string{"B", "KB", "MB", "GB", "TB"}
	value := math.Max(numBytes, 0)
	for i, unit := range units {
		if value < 1024 || i == len(units)-1 {
			if unit == "B" {
				return fmt.Sprintf("%d %s", int64(value), unit)
			}
			return trimZeros(value) + " " + unit
		}
		value /= 1024
	}
	return trimZeros(value) + " TB"
}

// trimZeros rounds to two decimals and drops trailing zeros, keeping at least
// one decimal place so sizes still read as measurements.
func trimZeros(value float64) string {
	rounded := math.Round(value*100) / 100
	text := strconv.FormatFloat(rounded, 'f', -1, 64)
	if !strings.Contains(text, ".") {
		text += ".0"
	}
	return text
}

// FormatDuration renders seconds as HH:MM:SS (or MM:SS under an hour).
func FormatDuration(seconds float64) string {
	if math.IsNaN(seconds) || math.IsInf(seconds, 0) || seconds < 0 {
		return "--"
	}
	total := int64(seconds)
	h := total / 3600
	m := (total % 3600) / 60
	s := total % 60
	if h > 0 {
		return fmt.Sprintf("%02d:%02d:%02d", h, m, s)
	}
	return fmt.Sprintf("%02d:%02d", m, s)
}

func clip(text string, maxLen int) string {
	text = strings.ReplaceAll(strings.TrimSpace(text), "\n", " ")
	runes := []rune(text)
	if len(runes) <= maxLen {
		return text
	}
	return string(runes[:maxLen-3]) + "..."
}

// Clone tracks the state of an in-flight clone. It is owned by a single
// goroutine: the one running the clone.
type Clone struct {
	TaskName    string
	TotalBytes  int64
	TotalFiles  int64
	CopiedBytes int64
	CopiedFiles int64
	StartedAt   time.Time
	Status      string
}

// NewClone creates a Clone for the named task.
func NewClone(taskName string) *Clone {
	return &Clone{
		TaskName:  taskName,
		StartedAt: time.Now(),
		Status:    "Initializing",
	}
}

// Percentage is byte-based when a total size is known, else file-count based.
func (c *Clone) Percentage() float64 {
	if c.TotalBytes > 0 {
		return math.Min(100, float64(c.CopiedBytes)/float64(c.TotalBytes)*100)
	}
	if c.TotalFiles > 0 {
		return math.Min(100, float64(c.CopiedFiles)/float64(c.TotalFiles)*100)
	}
	return 0
}

// Elapsed is clamped away from zero so it is always safe to divide by.
func (c *Clone) Elapsed() float64 {
	return math.Max(0.001, time.Since(c.StartedAt).Seconds())
}

// SpeedBPS is the average copy speed so far, in bytes per second.
func (c *Clone) SpeedBPS() float64 {
	if c.CopiedBytes <= 0 {
		return 0
	}
	return float64(c.CopiedBytes) / c.Elapsed()
}

// ETASeconds returns NaN when no meaningful estimate exists yet.
func (c *Clone) ETASeconds() float64 {
	if c.TotalBytes <= 0 {
		return math.NaN()
	}
	remaining := c.TotalBytes - c.CopiedBytes
	if remaining < 0 {
		remaining = 0
	}
	speed := c.SpeedBPS()
	if speed <= 0 {
		return math.NaN()
	}
	return float64(remaining) / speed
}

// Bar renders a fixed-width unicode progress bar.
func (c *Clone) Bar(width int) string {
	pct := math.Max(0, math.Min(100, c.Percentage()))
	filled := int(pct / 100 * float64(width))
	return "[" + strings.Repeat("▓", filled) + strings.Repeat("░", width-filled) + "]"
}

// Message renders the live progress message.
func (c *Clone) Message() string {
	totalSize := "Unknown"
	if c.TotalBytes != 0 {
		totalSize = FormatBytes(float64(c.TotalBytes))
	}

	return "<u><b>CLONING</b></u>\n" +
		fmt.Sprintf("<code>%s</code>\n", html.EscapeString(clip(c.TaskName, 90))) +
		fmt.Sprintf("<code>%s %.2f%%</code>\n", html.EscapeString(c.Bar(12)), c.Percentage()) +
		fmt.Sprintf("<code>%s</code> <code>/</code> <code>%s</code>\n",
			html.EscapeString(FormatBytes(float64(c.CopiedBytes))), html.EscapeString(totalSize)) +
		fmt.Sprintf("<code>↓%s | ETA : %s</code>",
			html.EscapeString(FormatBytes(c.SpeedBPS())+"/s"), html.EscapeString(FormatDuration(c.ETASeconds())))
}

// CompletionMessage renders the final message for a finished clone.
func (c *Clone) CompletionMessage(outputName, outputURL string) string {
	return "<u><b>CLONED</b></u>\n" +
		fmt.Sprintf("<code>%s</code>\n", html.EscapeString(outputName)) +
		fmt.Sprintf("<b>%s •</b> ", html.EscapeString(FormatBytes(float64(c.CopiedBytes)))) +
		fmt.Sprintf("<a href=\"%s\"><b>DL</b></a>", html.EscapeString(outputURL))
}

// Throttle rate-limits progress edits so Telegram does not throttle us instead.
type Throttle struct {
	MinInterval  time.Duration
	lastUpdateAt time.Time
}

// NewThrottle creates a Throttle with the given minimum edit interval.
func NewThrottle(minInterval time.Duration) *Throttle {
	return &Throttle{MinInterval: minInterval}
}

// ShouldUpdate reports whether enough time has passed since the last edit.
func (t *Throttle) ShouldUpdate(force bool) bool {
	now := time.Now()
	if force || now.Sub(t.lastUpdateAt) >= t.MinInterval {
		t.lastUpdateAt = now
		return true
	}
	return false
}
